backend/hidrive/hidrive.go

```diff
@@ -0,0 +1,1021 @@
+	modTime     time.Time // modification time of the object
+	id          string    // ID of the object
+	hash        string    // content-hash of the object
+}
+
+// ZeroDelayCalculator is a pacer.Calculator that never delays.
+type ZeroDelayCalculator struct{}
+
+// ------------------------------------------------------------
+
+// Calculate takes the current Pacer state and returns the sleep time after which
+// the next Pacer call will be done.
+func (c *ZeroDelayCalculator) Calculate(state pacer.State) time.Duration {
+	return 0
+}
+
+// Name returns the name of the remote (as passed into NewFs).
+func (f *Fs) Name() string {
+	return f.name
+}
+
+// Root returns the name of the remote (as passed into NewFs).
+func (f *Fs) Root() string {
+	return f.root
+}
+
+// String returns a string-representation of this Fs.
+func (f *Fs) String() string {
+	return fmt.Sprintf("HiDrive root '%s'", f.root)
+}
+
+// Precision returns the precision of this Fs.
+func (f *Fs) Precision() time.Duration {
+	return time.Second
+}
+
+// Hashes returns the supported hash sets.
+func (f *Fs) Hashes() hash.Set {
+	return hash.Set(hidrivehashType)
+}
+
+// Features returns the optional features of this Fs.
+func (f *Fs) Features() *fs.Features {
+	return f.features
+}
+
+// errorHandler parses a non 2xx error response into an error.
+func errorHandler(resp *http.Response) error {
+	// Decode error response.
+	errResponse := new(api.Error)
+	err := rest.DecodeJSON(resp, &errResponse)
+	if err != nil {
+		fs.Debugf(nil, "Couldn't decode error response: %v", err)
+	}
+	_, err = errResponse.Code.Int64()
+	if err != nil {
+		errResponse.Code = json.Number(strconv.Itoa(resp.StatusCode))
+	}
+	return errResponse
+}
+
+// NewFs creates a new file system from the path.
+func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, error) {
+	fs.Debugf(nil, "hidrive: creating new Fs.")
+	// Parse config into Options struct.
+	opt := new(Options)
+	err := configstruct.Set(m, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	// Clean root-prefix and root-path.
+	// NOTE: With the default-encoding "." and ".." will be encoded,
+	// but with custom encodings without encoder.EncodeDot
+	// "." and ".." will be interpreted as paths.
+	if opt.RootPrefix != "" {
+		opt.RootPrefix = path.Clean(opt.Enc.FromStandardPath(opt.RootPrefix))
+	}
+	root = path.Clean(opt.Enc.FromStandardPath(root))
+
+	client, ts, err := oauthutil.NewClient(ctx, name, m, oauthConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to configure HiDrive: %w", err)
+	}
+
+	f := &Fs{
+		name:    name,
+		root:    root,
+		opt:     *opt,
+		srv:     rest.NewClient(client).SetRoot(opt.EndpointAPI),
+		pacer:   fs.NewPacer(ctx, pacer.NewDefault(pacer.MinSleep(minSleep), pacer.MaxSleep(maxSleep), pacer.DecayConstant(decayConstant))),
+		retrier: pacer.New(pacer.RetriesOption(2), pacer.MaxConnectionsOption(-1), pacer.CalculatorOption(&ZeroDelayCalculator{})),
```

Comment:
Number of retries is traditionally controlled by `--low-level-retries` unless we have a strong reason not to, so you don't need to override this.

Reply:
This `retrier` is how I "abuse" rclone's pacer mechanism for something it was not originally designed for:
To repeat a fragment of code at most once, in case a necessary precondition is not fulfilled and the code needs to be repeated after that condition is fulfilled.
The `retrier` should never actually be used to repeat API-calls; that's what the above `pacer` object is for.

This is also what I briefly mentioned in [the discussion a while back](https://github.com/rclone/rclone/pull/5822#discussion_r750397613). As I mention there, the same effect could be accomplished through a for-loop or duplicated code.

I'll try to illustrate the motivation behind this using the `Copy` implementation:
We could first check if the parent-folders of the destination already exists, and create them accordingly before issuing the copy operation.
However we can expect that the parent-folders already exists most of the time, so this would be a bit inefficient.
Instead the design is as follows:
1. Try to copy the file.
2. If it fails because the parents are missing, create the parents and repeat step 1 *once*.

<details>
<summary>This could be implemented in the following ways (simplified):</summary>

1. duplicated code:
```golang
err = copyFile(...)
if err {
    if sourceIsMissing(...) {
      // fail. Abort copy-operation.
    }
    err = createParents(...)
    if err == nil {
        err = copyFile(...)
    }
}
```
2. for loop:
```golang
for attempt := 0; attempt < 2; attempt++ {
    err = copyFile(...)
    if err == nil {
        break
    } else {
        if sourceIsMissing(...) {
          // fail. Abort copy-operation.
        }
        if attempt == 0 {
            createParents(...)
            if err == nil {
                continue
            }
        }
    } 
}
```
3. `retrier`:
```golang
f.retrier.Call(func() (bool, error) {
    err = copyFile(...)
    if err {
        if sourceIsMissing(...) {
            return false, err
        }
        err = createParents(...)
        if err {
            return false, err
        }
        return true, err
    }
    return false, err
}
```
</details>

I decided to use the `retrier`-mechanism as it seemed more readable than the for loop I used before. Though, looking back at it now, it may have just been easiest to use duplicated code...

If you dislike this "abuse" of the pacer-mechanism I'll rewrite the 4 affected methods to use duplicated code instead.
Otherwise, maybe `retrier` needs a better name like `retryOnce` and a longer comment-block explaining the idea behind it?

Reply:
I see! Sorry careless reviewing on my part.

Yes I like the idea of renaming it `retryOnce` with a little comment - sounds perfect. You'll also get debug messages for retries which could well be useful.

We could also hoist `ZeroDelayCalculator` into `lib/pacers` - it looks like it could be useful elsewhere. Though it is only a few lines of code!

Reply:
Sounds good, I'll create a separate commit for `ZeroDelayCalculator` and move it to `lib/pacers` then!

Reply:
Great - thanks.